IQR Integration

danlamanna opened the pull request:
This is definitely a work in progress, but a usable implementation.

A few things to note:
- This branch includes PR #152 as it needed the helper function `solr_documents_from_field`
- This branch depends on girder/girder#1261
- User uploaded images can't be used for adjudication just yet
- The UI is ugly right now

The flow of how this works:
1) From any results view in ImageSpace an IQR session can be started from the top right
2) Once an IQR session is started - saving and searching based on an image is disabled, those buttons are replaced with an annotation widget that lets the user choose whether the image is positive or negative.
3) Once the user has selected enough of the images they want, they can hit the Refine button on the top right and they will be in a "refine" mode - from here they can only look through the refinement results and add more positive/negative results and further refine.

Left to do:
- [x] Clean up the UI
- [x] Make it clear the user is in refinement mode
- [x] Support removing images from positive/negative (i.e. clicking positive on a positive labeled image should make it neutral)
- [ ] Allow user uploads to be used for refinement (this is more on Paul)
- [ ] Allow saving/application of classifiers
- [x] Allow cancellation of refinement in initial stage
- [x] Call it "Interactive Refinement" instead of IQR for clarity
- [x] Gun type checkboxes do not work in refinement mode - make them work or hide them

@jeffbaumes Feel free to review if you want (noting the Girder dependency) and add any tasks you think need to be added.

danlamanna committed:
Initial implementation of server side endpoints for IQR

danlamanna committed:
Stop canceling rest requests when initializing views

danlamanna committed:
Properly load plugin JS after ImageSpace JS

danlamanna committed:
Stop fetching SearchView collection immediately

danlamanna committed:
Initial client side implementation of IQR

danlamanna committed:
Only find IQR Session once

danlamanna committed:
Fix IQR error on minimum exemplars

danlamanna committed:
Add notification for when a user is refining

danlamanna committed:
Allow user to deselect pos/neg annotations

danlamanna committed:
Cleanup UI and add refining feedback

chrismattmann commented:
LGTM @danlamanna as a first start!

Purg commented:
Also supported by the SMQTK service under the hood is a "/classify" endpoint that takes a session id and a list of descriptors by UUID (sha1), creates a binary classifier based on adjudication state and applies it to the descriptors specified, returning their positive class confidence scores. Requires a session with an existing adjudication state including at least one positive and one negative example.

It would be great to have a feature where you can apply an IQR session's classifier to an arbitrary page of results, marking session classifier confidence scores for the visible items (color, bar, percentile mark, etc.).

Purg commented:
Also its generally called "Interactive Query Refinement" I think. I get it confused a lot, too.

jeffbaumes commented:
Oops changed the name to "Interactive Refinement" - it "Interactive Query Refinement" fits in the space we have in the UI, I suppose that would be best.

Purg commented:
Dan wanted me to respond to the addition of the ``Should there be a "reinitialize" button ...`` check-box task.

In the demo UI, there were two buttons in order to control working index updating and refinement separately. The initial pass for the ImageSpace integration had me join those two steps into one, which is what's currently bundled under the "Refine" button there. When a user hits that button, the IQR service ("/refine" endpoint) will currently update the working index for any new positive examples provided, and then refine the working index. It is an open question as to whether two buttons should be provided to the user or not. To me, it seems like there is a trade off between intuitive functionality or a more optimized workflow requiring more specialized knowledge of what's going on systematically.

danlamanna committed:
Fix broken similarity searches

This was broken in ef30256 because we stopped fetching collections by
default. When looking for instantiations of the SearchView I did not
think to look for its use as a symbol.

jeffbaumes commented:
@Purg this sounds like a great approach. I didn't realize you are refining the working index as part of the refinement. For this app we definitely want to favor simplicity over full control of the process. I've removed by proposed task about this.

jeffbaumes commented:
I found this issue:

1. Perform a "*" search
2. Perform a CMU similarity search on the first image
3. Click the Start IQR Session button
4. The view presents the "*" search results for selecting Pos/Neg instead of the similarity search

danlamanna committed:
Clarify what IQR is

danlamanna committed:
Remove /refine endpoint

Instead the system will just consider things to be in refinement if
there is a parameter for an IQR session in the URL. This is so that any
existing searches remain unobstructed and just get tweaked by us.

danlamanna committed:
Navigate to similarity searches the same way we do other searches

danlamanna committed:
Fix permalinking for all search URLs, fix hiding of classification box

danlamanna committed:
Fix issues with iqr refining notice, implement quitting an IQR session

danlamanna commented:
Note to myself: a bug seems to have popped up where occasionally refinement results take the user to page 2 instead of page 1.

danlamanna committed:
Fix logical conflict with merge

danlamanna committed:
Fixing navigation issues

danlamanna committed:
Add to error log if there is an invalid mapping of descriptors to documents

danlamanna committed:
Document IQR env variable

jeffbaumes commented:
This has been out quite a while and it has been demonstrated to work well in a few domains. LGTM!

chrismattmann commented:
Woot!
